"borrow of possibly uninitialized variable" instead of "use of moved value"

NLL combined with a struct field assignment gives an error mentioning the use/borrow of an uninitialized variable instead of the use/borrow of a moved variable.

# Reproduction example

*Simplified code based on @Yatekii's question on IRC*
[Playground](https://play.rust-lang.org/?gist=3458f67aa4fb164d002ad105c6d6d497&version=nightly&mode=debug&edition=2015)
```rust
#![feature(nll)]

struct A {
    b: B,
}

#[derive(Clone)]
struct B;

fn foo(_: A) {}

fn bar(mut a: A) -> B {
    // Commenting the line below gives a move error instead of uninitialized
    // error.
    a.b = B;
    foo(a);
    a.b.clone()
}

fn main() {}
```

# Expected behavior

The compiler gives an error about using a moved value, similar to the following:

Without NLL:
```
error[E0382]: use of moved value: `a.b`
  --> src/main.rs:17:5
   |
16 |     foo(a);
   |         - value moved here
17 |     a.b.clone()
   |     ^^^ value used here after move
   |
   = note: move occurs because `a` has type `A`, which does not implement the `Copy` trait
```

With NLL but without the field assignment:
```
error[E0382]: borrow of moved value: `a.b`
  --> src/main.rs:17:5
   |
16 |     foo(a);
   |         - value moved here
17 |     a.b.clone()
   |     ^^^ value borrowed here after move
   |
   = note: move occurs because `a` has type `A`, which does not implement the `Copy` trait
```

# Actual behavior

The error mentions an uninitialized value instead of a moved value:

```
error[E0381]: borrow of possibly uninitialized variable: `a.b`
  --> src/main.rs:17:5
   |
17 |     a.b.clone()
   |     ^^^ use of possibly uninitialized `a.b`
```

this issue may caused by `Place` checking I guess:
https://github.com/rust-lang/rust/blob/71a1ef181683615f1c8da5efb3d8617c0b743009/src/librustc_mir/borrow_check/mod.rs#L1620
the comment reasons why the `uninitialized` lint emitted here: for place `a.b`, if `a` is moved, then the usage of projection `a.b` will be marked as `uninitialized`, let's take an experiment:
https://play.rust-lang.org/?gist=12d7880fc733e74b5bac20119465ada3&version=nightly&mode=debug&edition=2015
```rust
fn bar(mut a: A) -> B {
    a.b = B;
    foo(a);
    a.b
}
```
=> 
```
error[E0381]: use of possibly uninitialized variable: `a.b`
  --> src/main.rs:16:5
   |
16 |     a.b
   |     ^^^ use of possibly uninitialized `a.b`
```

==========

instead of using place `a.b`, we return `a`:
```rust
```rust
fn too(mut a: A) -> A {
    a.b = B;
    foo(a);
    a
}
```
=> 
```
error[E0382]: use of moved value: `a`
  --> src/main.rs:16:5
   |
15 |     foo(a);
   |         - value moved here
16 |     a
   |     ^ value used here after move
   |
   = note: move occurs because `a` has type `A`, which does not implement the `Copy` trait

```

--- a/src/librustc_mir/borrow_check/error_reporting.rs
+++ b/src/librustc_mir/borrow_check/error_reporting.rs
@@ -564,9 +564,20 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 // created by `StorageDead` and at the beginning
                 // of a function.
             } else {
+                // If we are found a use of a.b.c which was in error, then we want to look for
+                // moves not only of a.b.c but also a.b and a.
+                //
+                // Note that the moves data already includes "parent" paths, so we don't have to
+                // worry about the other case: that is, if there is a move of a.b.c, it is already
+                // marked as a move of a.b and a as well, so we will generate the correct errors
+                // there.
+                let mut mpis = vec![mpi];
+                let move_paths = &self.move_data.move_paths;
+                mpis.extend(move_paths[mpi].parents(move_paths));
+
                 for moi in &self.move_data.loc_map[l] {
                     debug!("report_use_of_moved_or_uninitialized: moi={:?}", moi);
-                    if self.move_data.moves[*moi].path == mpi {
+                    if mpis.contains(&self.move_data.moves[*moi].path) {
                         debug!("report_use_of_moved_or_uninitialized: found");
                         result.push(*moi);
 

--- a/src/librustc_mir/dataflow/move_paths/mod.rs
+++ b/src/librustc_mir/dataflow/move_paths/mod.rs
@@ -97,6 +97,20 @@ pub struct MovePath<'tcx> {
     pub place: Place<'tcx>,
 }
 
+impl<'tcx> MovePath<'tcx> {
+    pub fn parents(&self, move_paths: &IndexVec<MovePathIndex, MovePath>) -> Vec<MovePathIndex> {
+        let mut parents = Vec::new();
+
+        let mut curr_parent = self.parent;
+        while let Some(parent_mpi) = curr_parent {
+            parents.push(parent_mpi);
+            curr_parent = move_paths[parent_mpi].parent;
+        }
+
+        parents
+    }
+}
+
 impl<'tcx> fmt::Debug for MovePath<'tcx> {
     fn fmt(&self, w: &mut fmt::Formatter) -> fmt::Result {
         write!(w, "MovePath {{")?;

--- a/src/test/ui/nll/issue-52669.rs
+++ b/src/test/ui/nll/issue-52669.rs
@@ -1,0 +1,28 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(nll)]
+
+struct A {
+    b: B,
+}
+
+#[derive(Clone)]
+struct B;
+
+fn foo(_: A) {}
+
+fn bar(mut a: A) -> B {
+    a.b = B;
+    foo(a);
+    a.b.clone()
+}
+
+fn main() {}

--- a/src/test/ui/nll/issue-52669.stderr
+++ b/src/test/ui/nll/issue-52669.stderr
@@ -1,0 +1,13 @@
+error[E0382]: borrow of moved value: `a.b`
+  --> $DIR/issue-52669.rs:25:5
+   |
+LL |     foo(a);
+   |         - value moved here
+LL |     a.b.clone()
+   |     ^^^ value borrowed here after move
+   |
+   = note: move occurs because `a` has type `A`, which does not implement the `Copy` trait
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0382`.

--- a/src/test/ui/nll/move-subpaths-moves-root.rs
+++ b/src/test/ui/nll/move-subpaths-moves-root.rs
@@ -1,0 +1,17 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(nll)]
+
+fn main() {
+    let x = (vec![1, 2, 3], );
+    drop(x.0);
+    drop(x);
+}

--- a/src/test/ui/nll/move-subpaths-moves-root.stderr
+++ b/src/test/ui/nll/move-subpaths-moves-root.stderr
@@ -1,0 +1,13 @@
+error[E0382]: use of moved value: `x`
+  --> $DIR/move-subpaths-moves-root.rs:16:10
+   |
+LL |     drop(x.0);
+   |          --- value moved here
+LL |     drop(x);
+   |          ^ value used here after move
+   |
+   = note: move occurs because `x.0` has type `std::vec::Vec<i32>`, which does not implement the `Copy` trait
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0382`.

